overlord/managers_test.go

```diff
@@ -3255,34 +3316,127 @@ type: base`
+	defer partition.ForceBootloader(nil)
+
+	restore := release.MockOnClassic(false)
+	defer restore()
+
+	mockServer := ms.mockStore(c)
+	defer mockServer.Close()
+
+	st := ms.o.State()
+	st.Lock()
+	defer st.Unlock()
+
+	si := &snap.SideInfo{RealName: "pc-kernel", SnapID: fakeSnapID("pc-kernel"), Revision: snap.R(1)}
+	snapstate.Set(st, "pc-kernel", &snapstate.SnapState{
+		Active:   true,
+		Sequence: []*snap.SideInfo{si},
+		Current:  snap.R(1),
+		SnapType: "kernel",
+	})
+
+	const kernelYaml = `name: pc-kernel
+type: kernel
+version: 2.0`
+	snapPath, _ := ms.makeStoreTestSnap(c, kernelYaml, "2")
+	ms.serveSnap(snapPath, "2")
+
+	ms.prereqSnapAssertions(c, map[string]interface{}{
+		"snap-name": "foo",
+	})
+	snapPath, _ = ms.makeStoreTestSnap(c, `{name: "foo", version: 1.0}`, "1")
+	ms.serveSnap(snapPath, "1")
 
+	// create/set custom model assertion
+	model := makeModelAssertion(c, ms.storeSigning.SigningDB, map[string]interface{}{
+		"authority-id": "can0nical",
+		"brand-id":     "can0nical",
+	})
 	// setup model assertion
 	auth.SetDevice(st, &auth.DeviceState{
-		Brand: "my-brand",
+		Brand: "can0nical",
 		Model: "my-model",
 	})
-	err = assertstate.Add(st, model)
+	err := assertstate.Add(st, model)
 	c.Assert(err, IsNil)
 
 	// create a new model
-	newModel := makeModelAssertion(c, brandSigning, map[string]interface{}{
-		"base":     "core18",
-		"revision": "1",
+	newModel := makeModelAssertion(c, ms.storeSigning.SigningDB, map[string]interface{}{
+		"authority-id":   "can0nical",
+		"brand-id":       "can0nical",
+		"kernel":         "pc-kernel=18",
+		"revision":       "1",
+		"required-snaps": []interface{}{"foo"},
 	})
 
 	tss, err := devicestate.Remodel(st, newModel)
-	c.Assert(err, ErrorMatches, "cannot remodel to different bases yet")
-	c.Assert(tss, HasLen, 0)
+	c.Assert(err, IsNil)
+	chg := st.NewChange("remodel", "...")
+	c.Check(tss, HasLen, 3)
+	chg.AddAll(tss[0])
+	chg.AddAll(tss[1])
+	chg.AddAll(tss[2])
+
+	st.Unlock()
+	err = ms.o.Settle(settleTimeout)
+	st.Lock()
+	c.Assert(err, IsNil)
+
+	// system waits for a restart because of the new kernel
+	t := findKind(chg, "auto-connect")
+	c.Assert(t, NotNil)
+	c.Assert(t.Status(), Equals, state.DoingStatus)
+
+	// simulate successful restart happened
+	state.MockRestarting(st, state.RestartUnset)
+
+	// continue
+	st.Unlock()
+	err = ms.o.Settle(settleTimeout)
+	st.Lock()
+	c.Assert(err, IsNil)
+
+	c.Assert(chg.Status(), Equals, state.DoneStatus, Commentf("upgrade-snap change failed with: %v", chg.Err()))
+
+	// ensure tasks were run in the right order
+	tasks := chg.Tasks()
+	sort.Sort(byReadyTime(tasks))
+
+	// first all downloads/checks in sequential order
+	var i int
+	i += validateDownloadCheckTasks(c, tasks[i:], "pc-kernel", "2", "18")
+	i += validateDownloadCheckTasks(c, tasks[i:], "foo", "1", "stable")
+
+	// then all installs in sequential order
+	i += validateRefreshTasks(c, tasks[i:], "pc-kernel", "2")
+	validateInstallTasks(c, tasks[i:], "foo", "1")
```

Comment:
shouldn't we check that there are not left over tasks?  maybe that applies to previous tests as well